In [2]:
import h5py
import numpy as np

In [3]:
def survey(i):
    print('image_id={} has {} objects'.format(i, ((objects['features'][i] != 0.0).sum(axis=1) != 0).sum()))

# survey(0)
# survey(1)
# survey(2)

# Objects_gqa_info.json
{"**2370799**": {"width": 500, "objectsNum": 24, "idx": 4550, "height": 333, "file": 9}, "2370798": {"width": 333, "objectsNum": 13, "idx": 9008, "height": 500, "file": 1}, "**2370793**": {"width": 500, "objectsNum": 17, "idx": 8130, "height": 375, "file": 7}

So the bolded numbers are **image ids** 

take this for example 

    {"2370799": {"width": 500, "objectsNum": 24, "idx": 4550, "height": 333, "file": 9}

this says that that we have 24 objects in the image=2370799 stored in the index=4550 in the 9th object h5 file.

Hence we can access this image's object features as follows 

In [4]:
datapath = '../data/objects/gqa_objects_9.h5'
objects = h5py.File(datapath, 'r')
features = objects['features'][4550][:24]
print('the feature are ', features)
bbs = objects['bboxes'][4550][:24]
print('The bounding boxes are', bbs)

the feature are  [[0.0000000e+00 0.0000000e+00 6.0011845e+00 ... 6.8882656e+00
  0.0000000e+00 0.0000000e+00]
 [4.9177136e-02 8.9633679e-01 1.6788861e+00 ... 1.2332850e+00
  0.0000000e+00 0.0000000e+00]
 [0.0000000e+00 0.0000000e+00 8.6887451e-03 ... 0.0000000e+00
  0.0000000e+00 1.6653609e+00]
 ...
 [4.1230787e-03 1.2189344e-03 1.2778348e-01 ... 9.6291928e-03
  0.0000000e+00 0.0000000e+00]
 [0.0000000e+00 5.7812691e-02 1.0047505e+00 ... 8.4616970e-03
  0.0000000e+00 4.1550022e-02]
 [0.0000000e+00 0.0000000e+00 2.1322658e-02 ... 0.0000000e+00
  0.0000000e+00 3.7822976e+00]]
The bounding boxes are [[153.57312  174.32625  209.50914  230.22238 ]
 [ 10.982235 196.58748  499.5      332.445   ]
 [119.395645  86.49263  154.43341  108.92558 ]
 [  0.       134.84871  287.2134   332.445   ]
 [ 65.90839  124.3244   230.60796  232.38704 ]
 [  0.         0.       282.7023   181.075   ]
 [275.50995    0.       499.5      284.26956 ]
 [ 47.61806  173.83682   97.93482  226.47842 ]
 [ 55.976    159.610

# relations to scene graph 

I think your confusion arose from the existence of unique “object-ids” that exists in the scene graphs. Now the questions is then how do we relate the scene graph object ids to the object features in the h5py file?

Take this scene graph for image_id =  2329478

     {'width': 500, 
     'objects': {
        '4673973': {'name': 'pole', 'h': 14, 'relations': [], 'w': 499, 'attributes': [], 'y': 311, 'x': 1}, 
        '4673972': {'name': 'bird', 'h': 87, 'relations': [{'object': '4673976', 'name': 'to the left of'}, {'object': '4673973', 'name': 'on'}], 'w': 53, 'attributes': [], 'y': 253, 'x': 231}, 
        '4673984': {'name': 'building', 'h': 374, 'relations': [], 'w': 500, 'attributes': [], 'y': 0, 'x': 0}, 
        '4673976': {'name': 'window', 'h': 329, 'relations': [{'object': '4673972', 'name': 'to the right of'}], 'w': 115, 'attributes': [], 'y': 16, 'x': 288}}, 'height': 375}


In [7]:
import json
sG = json.load(open('../data/sceneGraphs/train_sceneGraphs.json', 'r'))
objects_info = json.load(open('../data/objects/gqa_objects_info.json', 'r'))

In [8]:
image_id = 2329478
print('Scene graph for image_id={}\n {}\n'.format(image_id, sG[str(image_id)]))
print('Object info for image_id={}\n {}\n'.format(image_id, objects_info[str(image_id)]))

print('this image has these objects')
print(list(sG[str(image_id)]['objects'].keys()))
object_id = list(sG[str(image_id)]['objects'].keys())[0]
print('taking the first one.. the object id={} from the image_id={}\n'.format(object_id, image_id))

print('looking at it from the ojbect info.json.. for the image_id={}'.format(image_id))
objects_info[str(image_id)]

Scene graph for image_id=2329478
 {'width': 500, 'objects': {'4673973': {'name': 'pole', 'h': 14, 'relations': [], 'w': 499, 'attributes': [], 'y': 311, 'x': 1}, '4673972': {'name': 'bird', 'h': 87, 'relations': [{'object': '4673976', 'name': 'to the left of'}, {'object': '4673973', 'name': 'on'}], 'w': 53, 'attributes': [], 'y': 253, 'x': 231}, '4673984': {'name': 'building', 'h': 374, 'relations': [], 'w': 500, 'attributes': [], 'y': 0, 'x': 0}, '4673976': {'name': 'window', 'h': 329, 'relations': [{'object': '4673972', 'name': 'to the right of'}], 'w': 115, 'attributes': [], 'y': 16, 'x': 288}}, 'height': 375}

Object info for image_id=2329478
 {'width': 500, 'objectsNum': 14, 'idx': 319, 'height': 375, 'file': 5}

this image has these objects
['4673973', '4673972', '4673984', '4673976']
taking the first one.. the object id=4673973 from the image_id=2329478

looking at it from the ojbect info.json.. for the image_id=2329478


{'width': 500, 'objectsNum': 14, 'idx': 319, 'height': 375, 'file': 5}

now the confusion is that we only have 4 objects in the scene graph - but we have 14 objects in the object_info.json.

This is because we take detect image regions from Faster-RCNN and encode that into our feaetures h5py file. 
 
These DO NOT equal to the objects in the scene graph.


We can see this from the fact that the bounding boxes are different

hence, the object ids from the scene graph are not mapped to the object features h5py - the former is curated, and the latter are just automatically saved from Faster-RCNN.


In [9]:
datapath = '../data/objects/gqa_objects_5.h5'
objects = h5py.File(datapath, 'r')
bbs = objects['bboxes'][319][:14]
print('The bounding boxes are', bbs)

The bounding boxes are [[  0.         0.       499.375    318.73648 ]
 [229.54271  258.07455  280.7074   335.98068 ]
 [  0.       118.08587  499.375    374.375   ]
 [451.82828   35.56199  499.375    302.3901  ]
 [413.39597    2.016573 441.09784  305.53387 ]
 [ 41.28254  302.68912  490.21878  357.61874 ]
 [  0.         0.       141.90135  372.6703  ]
 [235.53879  299.02878  262.34436  343.5903  ]
 [333.75916   13.192463 414.20456  305.25882 ]
 [110.23621  331.70242  366.84573  372.38654 ]
 [ 16.97216  310.1954   386.79456  329.725   ]
 [131.03316   38.11346  159.95183  310.7696  ]
 [134.67415   23.685493 264.26642  304.1112  ]
 [305.01465   10.377226 407.21924  285.8725  ]]


In [33]:
def IoU(boxA, boxB):
	# determine the (x, y)-coordinates of the intersection rectangle
	xA = max(boxA[0], boxB[0])
	yA = max(boxA[1], boxB[1])
	xB = min(boxA[2], boxB[2])
	yB = min(boxA[3], boxB[3])
	# compute the area of intersection rectangle
	interArea = max(0, xB - xA + 1) * max(0, yB - yA + 1)
	# compute the area of both the prediction and ground-truth
	# rectangles
	boxAArea = (boxA[2] - boxA[0] + 1) * (boxA[3] - boxA[1] + 1)
	boxBArea = (boxB[2] - boxB[0] + 1) * (boxB[3] - boxB[1] + 1)
	# compute the intersection over union by taking the intersection
	# area and dividing it by the sum of prediction + ground-truth
	# areas - the interesection area
	iou = interArea / float(boxAArea + boxBArea - interArea)
	# return the intersection over union value
	return iou


In [34]:
bbx = {key: (int(value['x']), int(value['y']), int(value['h']), int(value['w']))for key, value in sG[str(image_id)]['objects'].items()}

In [35]:
for key, boxA in bbx.items():
    for boxB in objects['bboxes'][319][:14]:
        print(IoU(boxA, boxB))

0.0007526266203100454
0.0
0.00690688207628473
0.0
0.0
0.0
0.015904259982554667
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.5259701206274886
0.021908788042302854
0.43829678505823716
0.0
0.0
0.09602410736413489
0.2842208348267271
0.0067431230530767385
0.06053515338298576
0.05715621845175822
0.03906803560039923
0.043579122097450185
0.19561955348175225
0.09819660188516466
0.02625194567333743
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.08263158962430221


In [36]:
bbx = {}
for k ,v in sG[str(image_id)]['objects'].items():
    bbx[k] = (int(v['x']), int(v['y']), int(v['w'])+ int(v['x']), int(v['h']) + int(v['y']))

bbx

{'4673973': (1, 311, 500, 325),
 '4673972': (231, 253, 284, 340),
 '4673984': (0, 0, 500, 374),
 '4673976': (288, 16, 403, 345)}